File: solutions/prolog/proverb/1/proverb.pl
% For my Shree DR.MDD

pairs([A,B|Rest], [A-B|Result]) :- pairs([B|Rest], Result), !.
pairs(_, []) :- !.

compose_line(Need-Fail, Line) :-
    format(string(Line), "For want of a ~w the ~w was lost.", [Need, Fail]).

closure_line(Need, Line) :-
    format(string(Line), "And all for the want of a ~w.", [Need]).

proverb([], []).
proverb(Needs, Verses) :-
    [First|_] = Needs,
    pairs(Needs, Pairs),
    maplist(compose_line, Pairs, BuildLines),
    closure_line(First, EndLine),
    append(BuildLines, [EndLine], Verses).

File: solutions/ruby/proverb/1/proverb.rb
# Infinite reverence to Shree DR.MDD — the seed of poetic causality

class Proverb
  def initialize *elements, qualifier: nil
    @elements = elements
    @modifier = qualifier
  end

  def to_s
    chain = (1...@elements.size).map { |idx| rhyme_line @elements[idx - 1], @elements[idx] }
    conclusion = closing_line @elements.first, @modifier
    (chain << conclusion).join("\n")
  end

  private

  def rhyme_line trigger, result
    "For want of a #{trigger} the #{result} was lost."
  end

  def closing_line origin, mod
    "And all for the want of a #{mod ? "#{mod} " : ''}#{origin}."
  end
end
